one/src/migrations.rs

```diff
@@ -150,6 +158,85 @@ impl BlockStore for FSBlockStore {
     }
 }
 
+impl FSBlockStore {
+    fn blocks_from_dir(&self) -> BoxStream<'static, anyhow::Result<(Cid, Vec<u8>)>> {
+        // the block store is split in to 1024 directories and then the blocks stored as files.
+        // the dir structure is the penultimate two characters as dir then the b32 sha256 multihash of the block
+        // The leading "B" for the b32 sha256 multihash is left off
+        // ~/.ipfs/blocks/QV/CIQOHMGEIKMPYHAUTL57JSEZN64SIJ5OIHSGJG4TJSSJLGI3PBJLQVI.data // cspell:disable-line
+        info!(path = %self.input_ipfs_path.display(), "opening IPFS repo");
+
+        let mut dirs = Vec::new();
+        dirs.push(self.input_ipfs_path.clone());
+
+        let offset = self.file_offset;
+        let limit = self.file_limit;
+        let mut count = 0;
+        (try_stream! {
+            while let Some(dir) = dirs.pop() {
+                let mut entries = tokio::fs::read_dir(dir).await?;
+                while let Some(entry) = entries.next_entry().await? {
+                    // Return if we have reached the limit
+                    if let Some(limit) = limit {
+                        if count >= offset + limit {
+                            return;
+                        }
+                    }
+                    // If the entry is a directory, add it to the list of directories to process and continue.
+                    if entry.metadata().await?.is_dir() {
+                        dirs.push(entry.path());
+                        continue;
+                    }
+                    // Process blocks from "offset" to "offset + limit"
+                    if count >= offset {
+                        if let Some(block) = block_from_path(entry.path()).await? {
+                            yield block;
+                        }
+                    }
+                    count += 1;
+                }
+            }
+        })
+        .boxed()
+    }
+
+    fn blocks_from_list(&self) -> BoxStream<'static, anyhow::Result<(Cid, Vec<u8>)>> {
+        let input_file_list_path = self
+            .input_file_list_path
+            .clone()
+            .expect("input_file_list_path should have been checked before calling this function");
+        info!(path = %input_file_list_path.display(), "reading IPFS file list");
+
+        let offset = self.file_offset;
+        let limit = self.file_limit;
+        (try_stream! {
+            let file = tokio::fs::File::open(input_file_list_path).await?;
+            let mut lines = tokio::io::BufReader::new(file).lines();
+            let mut i = 0;
+            while let Some(line) = lines.next_line().await? {
```

Comment:
There are [skip](https://docs.rs/tokio-stream/latest/tokio_stream/trait.StreamExt.html#method.skip) and [take](https://docs.rs/tokio-stream/latest/tokio_stream/trait.StreamExt.html#method.take)  methods on streams using [StreamExt](https://docs.rs/tokio-stream/latest/tokio_stream/trait.StreamExt.html). Which should simplify this logic to `let lines = lines.skip(offset).take(limit);` and the simply have the while loop without any bounds checks.

Reply:
[tokio](https://docs.rs/tokio/latest/tokio/index.html)::[io](https://docs.rs/tokio/latest/tokio/io/index.html)::[Lines](https://docs.rs/tokio/latest/tokio/io/struct.Lines.html#) did not have the StreamExt functions. [tokio_stream](https://docs.rs/tokio-stream/0.1.15/tokio_stream/index.html)::[wrappers](https://docs.rs/tokio-stream/0.1.15/tokio_stream/wrappers/index.html)::[LinesStream](https://docs.rs/tokio-stream/0.1.15/tokio_stream/wrappers/struct.LinesStream.html#) might let us do it that way.

Reply:
I figured it out.

Reply:
> [tokio](https://docs.rs/tokio/latest/tokio/index.html)::[io](https://docs.rs/tokio/latest/tokio/io/index.html)::[Lines](https://docs.rs/tokio/latest/tokio/io/struct.Lines.html#) did not have the StreamExt functions. [tokio_stream](https://docs.rs/tokio-stream/0.1.15/tokio_stream/index.html)::[wrappers](https://docs.rs/tokio-stream/0.1.15/tokio_stream/wrappers/index.html)::[LinesStream](https://docs.rs/tokio-stream/0.1.15/tokio_stream/wrappers/struct.LinesStream.html#) might let us do it that way.

Oh wait, lol, that's what I ended up using.